migrations/type_value/migration_test.go

```diff
@@ -0,0 +1,472 @@
+	"github.com/onflow/cadence/runtime"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/tests/runtime_utils"
+	"github.com/onflow/cadence/runtime/tests/utils"
+)
+
+var _ migrations.Reporter = &testReporter{}
+
+type testReporter struct {
+	migratedPaths map[interpreter.AddressPath]struct{}
+}
+
+func newTestReporter() *testReporter {
+	return &testReporter{
+		migratedPaths: map[interpreter.AddressPath]struct{}{},
+	}
+}
+
+func (t *testReporter) Report(
+	addressPath interpreter.AddressPath,
+	_ string,
+) {
+	t.migratedPaths[addressPath] = struct{}{}
+}
+
+func TestTypeValueMigration(t *testing.T) {
+	t.Parallel()
+
+	account := common.Address{0x42}
+	pathDomain := common.PathDomainPublic
+
+	const stringType = interpreter.PrimitiveStaticTypeString
+
+	const fooBarQualifiedIdentifier = "Foo.Bar"
+	const fooBazQualifiedIdentifier = "Foo.Baz"
+
+	fooAddressLocation := common.NewAddressLocation(nil, account, "Foo")
+
+	newIntersectionStaticTypeWithoutInterfaces := func() *interpreter.IntersectionStaticType {
+		return interpreter.NewIntersectionStaticType(
+			nil,
+			[]*interpreter.InterfaceStaticType{},
+		)
+	}
+
+	newIntersectionStaticTypeWithOneInterface := func() *interpreter.IntersectionStaticType {
+		return interpreter.NewIntersectionStaticType(
+			nil,
+			[]*interpreter.InterfaceStaticType{
+				interpreter.NewInterfaceStaticType(
+					nil,
+					nil,
+					fooBarQualifiedIdentifier,
+					common.NewTypeIDFromQualifiedName(
+						nil,
+						fooAddressLocation,
+						fooBarQualifiedIdentifier,
+					),
+				),
+			},
+		)
+	}
+
+	newIntersectionStaticTypeWithTwoInterfaces := func() *interpreter.IntersectionStaticType {
+		return interpreter.NewIntersectionStaticType(
+			nil,
+			[]*interpreter.InterfaceStaticType{
+				interpreter.NewInterfaceStaticType(
+					nil,
+					nil,
+					fooBarQualifiedIdentifier,
+					common.NewTypeIDFromQualifiedName(
+						nil,
+						fooAddressLocation,
+						fooBarQualifiedIdentifier,
+					),
+				),
+				interpreter.NewInterfaceStaticType(
+					nil,
+					nil,
+					fooBazQualifiedIdentifier,
+					common.NewTypeIDFromQualifiedName(
+						nil,
+						fooAddressLocation,
+						fooBazQualifiedIdentifier,
+					),
+				),
+			},
+		)
+	}
+
+	type testCase struct {
+		storedType   interpreter.StaticType
+		expectedType interpreter.StaticType
+	}
+
+	testCases := map[string]testCase{
+		// base cases
+		"primitive": {
+			storedType:   stringType,
+			expectedType: nil,
+		},
+		"intersection_without_interfaces": {
+			storedType:   newIntersectionStaticTypeWithoutInterfaces(),
+			expectedType: nil,
+		},
+		"intersection_with_one_interface": {
+			storedType:   newIntersectionStaticTypeWithOneInterface(),
+			expectedType: nil,
+		},
+		"intersection_with_two_interfaces": {
+			storedType:   newIntersectionStaticTypeWithTwoInterfaces(),
+			expectedType: newIntersectionStaticTypeWithTwoInterfaces(),
+		},
```

Comment:
Is it possible to add a test that would change the interface order?
I guess that's impossible, because the initial storing of the value would order it?

Reply:
So I wrote a test case for this, storing a dictionary with a `Type` value which has the old type ID / hash, and then running the migration on it. 

This actually showed that the migration doesn't yet worked on actual old stored data. Started a thread in https://discord.com/channels/613813861610684416/1108479699732152503/1186414080731390054

Reply:
This is done now